File: materias 3 periodo ple/programacao orientada ao objeto/trabalhos praticos/tp-1/Codigo/Cpp/Administrador.cpp
#include <iostream>
#include <string>

#include "../Hpp/Administrador.hpp"

Administrador :: Administrador(string nome, string cpf, string rg, string endereco, string telefone, string email, string user, string senha, int id) : Funcionario(nome, cpf, rg, endereco, telefone, email, user, senha, id) {}

Administrador :: ~Administrador() {}

// Identificador
int Administrador :: idFunc = 1;

Vendedor Administrador :: criaVendedor() {
    // Forçamos o construtor para escrita dos campos desejados
    string user, senha;
    Pessoa newPessoa;
    cin >> newPessoa;
    cout << "User: ";
    getline(cin, user);
    cout << "Senha: ";
    getline(cin, senha);
    Vendedor newVendedor(newPessoa.getNome(), newPessoa.getCpf(), newPessoa.getRg(), newPessoa.getEndereco(), newPessoa.getTelefone(), newPessoa.getEmail(), user, senha);
    newVendedor.setId(idFunc);
    idFunc++;
    return newVendedor;
}

void Administrador :: addVendedor(Vendedor vendedor) {
    // Inserindo no fim da lista de vendedores
    this->vendedores.push_back(vendedor);
}

void Administrador :: removeVendedor(Vendedor vendedor) {
    // Removendo um vendedor de acordo com o id percorrido
    list<Vendedor>::iterator i = this->vendedores.begin();
    
    do {
        if(i->getId() == vendedor.getId())
            break;
        
        advance(i, 1);
    } while(i->getId() != vendedor.getId());

    this->vendedores.erase(i);
}

Vendedor Administrador :: getVendedorUser(string user) {
    // Obtendo um vendedor atraves do user passado.
    for(list<Vendedor>::iterator i = this->vendedores.begin(); i != this->vendedores.end(); i++) {
        if(user.compare(i->getUser()) == 0)
            return *i;
    } 
}

Vendedor Administrador :: getVendedorId(int id) {
    // Obtendo um vendedor atraves do id passado.
    for(list<Vendedor>::iterator i = this->vendedores.begin(); i != this->vendedores.end(); i++) {
        if(i->getId() == id)
            return *i;
    }
}

void Administrador :: setVendedor(Vendedor vendedor, int id) {
    for(list<Vendedor>::iterator i = this->vendedores.begin(); i != this->vendedores.end(); i++) {
        if(i->getId() == id){
            *i = vendedor;
            return;
        }
    }
}

Mecanico Administrador :: criaMecanico() {
    // Forçamos o construtor para escrita dos campos desejados
    string user, senha;
    Pessoa newPessoa;
    cin >> newPessoa;
    cout << "User: ";
    getline(cin, user);
    cout << "Senha: ";
    getline(cin, senha);
    Mecanico newMecanico(newPessoa.getNome(), newPessoa.getCpf(), newPessoa.getRg(), newPessoa.getEndereco(), newPessoa.getTelefone(), newPessoa.getEmail(), user, senha);
    newMecanico.setId(idFunc);
    idFunc++;
    return newMecanico;
}

void Administrador :: addMecanico(Mecanico mecanico) {
    // Inserindo no fim da lista de mecanicos
    this->mecanicos.push_back(mecanico);
}

void Administrador :: removeMecanico(Mecanico mecanico) {
    // Removendo um mecanico de acordo com o id percorrido
    list<Mecanico>::iterator i = this->mecanicos.begin();
    
    do {
        if(i->getId() == mecanico.getId())
            break;
        
        advance(i, 1);
    } while(i->getId() != mecanico.getId());

    this->mecanicos.erase(i);
}

Mecanico Administrador :: getMecanicoUser(string user) {
        // Obtendo um mecanico atraves do user passado.
    for(list<Mecanico>::iterator i = this->mecanicos.begin(); i != this->mecanicos.end(); i++) {
        if(user.compare(i->getUser()) == 0)
            return *i;
    } 
}

Mecanico Administrador :: getMecanicoId(int id) {
    // Obtendo um mecanico atraves do id passado.
    for(list<Mecanico>::iterator i = this->mecanicos.begin(); i != this->mecanicos.end(); i++) {
        if(i->getId() == id)
            return *i;
    }
}

void Administrador :: setMecanico(Mecanico mecanico, int id) {
    for(list<Mecanico>::iterator i = this->mecanicos.begin(); i != this->mecanicos.end(); i++) {
        if(i->getId() == id){
            *i = mecanico;
            return;
        }
    }
}

void Administrador :: printaVendedores() {
    // Printa em tela todos os vendedores
    cout << "\nId " << "Nome"  << "\n\n"; 
    for( list<Vendedor>::iterator i = this->vendedores.begin(); i != this->vendedores.end(); i++ ) {
        cout << i->getId() << " " << i->getNome() << "\n";
        cout << i->getUser() << " " << i->getNome() << "\n";
        cout << i->getSenha() << " " << i->getNome() << "\n";
        cout << "\n";
    }
}

void Administrador :: printaVendedor(Vendedor vendedor) {
    // Printa em tela um vendedor de acordo com o vendedor passado.
    cout << "\n";
    for( list<Vendedor>::iterator i = this->vendedores.begin(); i != this->vendedores.end(); i++ ) {
        if(i->getId() == vendedor.getId()){
            cout << "Nome      : " << i->getNome() << endl;
            cout << "Telefone  : " << i->getTelefone() << endl;
            cout << "Id        : " << i->getId() << endl;
            cout << "User      : " << i->getUser() << endl;
            cout << "Senha     : " << i->getSenha() << endl;
            cout << "\n";
            break;
        }
    }
}

void Administrador :: printaMecanicos() {
    // Printa em tela todos os mecanicos
    cout << "\nId " << "Nome"  << "\n\n"; 
    for( list<Mecanico>::iterator i = this->mecanicos.begin(); i != this->mecanicos.end(); i++ ) {
        cout << i->getId() << " " << i->getNome() << "\n";
        cout << "\n";
    }
}

void Administrador :: printaMecanico(Mecanico mecanico) {
    // Printa em tela um mecanico de acordo com o mecanico passado.
    cout << "\n";
    for( list<Mecanico>::iterator i = this->mecanicos.begin(); i != this->mecanicos.end(); i++ ) {
        if(i->getId() == mecanico.getId()){
            cout << "Nome      : " << i->getNome() << endl;
            cout << "Telefone  : " << i->getTelefone() << endl;
            cout << "Id        : " << i->getId() << endl;
            cout << "User      : " << i->getUser() << endl;
            cout << "Senha     : " << i->getSenha() << endl;
            cout << "\n";
            break;
        }
    }
}

int Administrador :: verificaUser(string user, string senha){
    // Realiza a validação de acesso comparando o padrao de entrada das categorias de funcionários, identificação e senha.
    string id;
    string userSenha;
    if(user[0] == 'v' && user[1] == 'e' && user[2] == 'n' && user[3] == 'd' && user[4] == 'e'){
        for(list<Vendedor>::iterator i = this->vendedores.begin(); i!= this->vendedores.end(); i++){
            id = i->getUser();
            if(user[5] == id[5] && user[6] == id[6] && user[7] == id[7] && user[8] == id[8]){
                userSenha = i->getSenha();                  
                if(senha == userSenha){                    
                    return 1;
                }
                else{
                    cout<<"Senha invalida"<<endl;
                    return 0;
                }
            }
        }
    }

    else if(user[0] == 'm' && user[1] == 'e' && user[2] == 'c' && user[3] == 'a' && user[4] == 'n'){
        string id;
        string userSenha;
        for(list<Mecanico> :: iterator i = this->mecanicos.begin(); i!= this->mecanicos.end(); i++){
            id = i->getUser();
            if(user[5] == id[5] && user[6] == id[6] && user[7] == id[7] && user[8] == id[8]){
                userSenha = i->getSenha();
                if(senha == userSenha){
                    return 2;
                }
                else{
                    cout<<"Senha invalida"<<endl;
                    return 0;
                }
                
            }
        }
    }
    else if(user[0] == 'a' && user[1] == 'd' && user[2] == 'm' && user[3] == 'i' && user[4] == 'n'){
        if(senha == "admin"){
            return 3; 
        }
        else{
            cout<<"Senha invalida"<<endl;
            return 0;
        }
    }
    else{
        cout<<"Usuario invalido"<<endl;
        return 0;
    }
    
}

File: materias 3 periodo ple/programacao orientada ao objeto/trabalhos praticos/tp-2/SistemaOficina/src/sistemaoficina/Funcionarios/Administrador.java
package sistemaoficina.Funcionarios;
import java.util.List;
import java.util.ArrayList;

public class Administrador extends Funcionario {
    private static int idFunc = 1001;
    private List<Vendedor> vendedores;
    private List<Mecanico> mecanicos;

    /**
    * Construtor da classe Administrador com todos os parametros preenchidos
    */
    public Administrador(String nome, String cpf, String rg, String endereco, String telefone, String email) {
        super(nome, cpf, rg, endereco, telefone, email, "admin", "admin");
        setId(0);
        vendedores = new ArrayList<>();
        mecanicos = new ArrayList<>();
    }

    public Administrador() {}

    /**
    * Retorna o campo vendedores da classe Administrador
    * @return vendedores - Campo que armazena os vendedores cadastrados
    */
    public List<Vendedor> getListaVendedores() {
        return vendedores;
    }
    
    /**
    * Retorna o campo mecanicos da classe Administrador
    * @return mecanicos - Campo que armazena os mecanicos cadastrados
    */
    public List<Mecanico> getListaMecanicos() {
        return mecanicos;
    }
    
    //Verifica se um funcionário existe de acordo com o ID
    public int existFunc(int op, int id){
        //Vendedor 1
        if(op == 1){
            for(Vendedor vend : vendedores) {
                if(vend.getId() == id){
                    return 1;
                }
                else {
                    return 0;
                }
            }
        }
        //Mecanico 2
        else{
            for(Mecanico meca : mecanicos) {
                if(meca.getId() == id){
                    return 1;
                }
                else {
                    return 0;
                }
            }
        }
        return 0;
    }

    /**
    * Retorna o campo idFunc da classe Administrador
    * @return idFunc - Campo responsavel pela contagem de funcionarios e para atribuir Id's unicos
    */
    public int getIdFunc() {
        return idFunc;
    }

    /**
    * Incrementa 1 ao campo idFunc
    */
    public static void setIdFunc() {
        idFunc++;
    }

    /**
    * Adiciona o vendedor passado nos parametros a lista de vendedores
    * @param vendedor - Objeto que será adicionado
    * @return Caso seja true o vendedor foi adicionado com sucesso,
    * caso contrario o vendedor em questão já foi adicionado em algum momento
    */
    public boolean addVendedor(Vendedor vendedor) {
        if(!vendedores.contains(vendedor)) {
            vendedor.setId(idFunc);
            vendedores.add(vendedor);
            Administrador.idFunc++;
            return true;
        }
        
        return false;
    }

    /**
    * Remove o vendedor passado nos parametros da lista de vendedores
    * @param vendedor - Objeto que será removido
    */
    public void removeVendedor(Vendedor vendedor) {
        if(!vendedores.isEmpty()) {
            vendedores.remove(vendedor);
        }
    }

    /**
    * Retorna da lista de vendedores o vendedor que tenha o id igual ao passado nos parametros
    * @param id - Referente ao vendedor que será retornado
    * @return Caso o vendedor seja encontrado retorna o vendedor, caso contrario
    * será retornado null
    */
    public Vendedor getVendedor(int id) {
        for(Vendedor vend : vendedores) {
            if(vend.getId() == id)
                return vend;
        }

        return null;
    }
    
    /**
    * Retorna da lista de vendedores o vendedor que tenha o user igual ao passado nos parametros
    * @param user - Referente ao vendedor que será retornado
    * @return Caso o vendedor seja encontrado retorna o vendedor, caso contrario
    * será retornado null
    */
    public Vendedor getVendedorUser(String user) {
        for(Vendedor vend : vendedores) {
            if(user.equals(vend.getUser()))
                return vend;
        }
        
        return null;
    }

    /**
    * Atualiza o vendedor já contido na lista, caso não exista o vendedor na lista, não acontecerá nada
    * @param vendedor - Objeto com os campos atualizados
    */
    public void setVendedor(Vendedor vendedor) {
        for(Vendedor vend : vendedores) {
            if(vend.equals(vendedor)){
                vend.setAtributos(vend, vendedor);
            }
        }
    }
    
    /**
    * Adiciona o mecanico passado nos parametros a lista de mecanicos
    * @param mecanico - Objeto que será adicionado
    * @return Caso seja true o mecanico foi adicionado com sucesso,
    * caso contrario o mecanico em questão já foi adicionado em algum momento
    */
    public boolean addMecanico(Mecanico mecanico) {
        if(!mecanicos.contains(mecanico)) {
            mecanico.setId(idFunc);
            mecanicos.add(mecanico);
            Administrador.idFunc++;
            return true;
        }
        
        return false;
    }
    
    /**
    * Remove o mecanico passado nos parametros da lista de mecanicos
    * @param mecanico - Objeto que será removido
    */
    public void removeMecanico(Mecanico mecanico) {
        if(!mecanicos.isEmpty()) {
            mecanicos.remove(mecanico);
        }
    }

    /**
    * Retorna da lista de mecanicos o mecanico que tenha o id igual ao passado nos parametros
    * @param id - Referente ao mecanico que será retornado
    * @return Caso o mecanico seja encontrado retorna o mecanico, caso contrario
    * será retornado null
    */
    public Mecanico getMecanico(int id) {
        for(Mecanico meca : mecanicos) {
            if(meca.getId() == id)
                return meca;
        }

        return null;
    }
    
    /**
    * Retorna da lista de mecanicos o mecanico que tenha o user igual ao passado nos parametros
    * @param user - Referente ao mecanico que será retornado
    * @return Caso o mecanico seja encontrado retorna o mecanico, caso contrario
    * será retornado null
    */
    public Mecanico getMecanicoUser(String user) {
        for(Mecanico meca : mecanicos) {
            if(user.equals(meca.getUser()))
                return meca;
        }
        
        return null;
    }
    
    /**
    * Atualiza o mecanico já contido na lista, caso não exista o mecanico na lista, não acontecerá nada
    * @param mecanico - Objeto com os campos atualizados
    */
    public void setMecanico(Mecanico mecanico) {
        for(Mecanico meca : mecanicos) {
            if(meca.equals(mecanico))
                meca.setAtributos(meca, mecanico);
        }
    }
    
    /**
    * Função responsavel pela verificação dos campos user e senha inseridos no login
    * @param user - User do usuario
    * @param senha - Senha do usuario
    * @return Caso o retorno seja 1 foi encontrado um vendedor, caso seja 2 foi encontrado um mecanico,
    * caso seja 3 foi encontrado um administrador e caso seja 0 não foi encontrado nenhum objeto
    */
    public int VerificaUsuario(String user, String senha) {
        if(user.contains("vende")) {
            String id = user.substring(5);
            int i = Integer.parseInt(id);
            
            Vendedor vend = getVendedor(i);
            if(vend != null) {
                if(senha.equals(vend.getSenha()))
                    return 1;
                
                else
                    return 0;
            }
            
            else
                return 0;
        }
        
        else if(user.contains("mecan")) {
            String id = user.substring(5);
            int i = Integer.parseInt(id);
            
            Mecanico meca = getMecanico(i);
            if(meca != null) {
                if(senha.equals(meca.getSenha()))
                    return 2;
                
                else
                    return 0;
            }
            
            else
                return 0;
        }
        
        else if(user.contains("admin")) {
            if(senha.equals("admin"))
                return 3;
            
            else
                return 0;
        }

        return 0;
    }

}
